{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d42495e-ab92-4e2e-bf50-a785e9b55b8c",
   "metadata": {},
   "source": [
    "# Binary classification with scikit-learn\n",
    "--------------------------\n",
    "### Contents:\n",
    "- **Background**\n",
    "- **Data importing/preprocessing - [10 points]**\n",
    "- **Feature selection and dimensionality reduction - [15 points]**\n",
    "- **SVM classifier - [20 points]**\n",
    "- **Logistic regression - [30 points]**\n",
    "- **Random forest - [25 points]**\n",
    "- **1 minute feedback - [Optional]**\n",
    "\n",
    "--------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbaf86a8-9c11-4f07-8048-58b890db34a3",
   "metadata": {},
   "source": [
    "### Section 1: Background \n",
    "In this assignment, your task is to build a binary classifier that can distinguish between two brain disorders - Alzheimer's (AD) and Progressive Supranuclear Palsy (PSP). We will be using gene expression data sourced by RNA sequencing (RNA-seq) the brain tissue of  participants. The study can be found at [<sup>1</sup>](#fn1).\n",
    "\n",
    "Using RNA sequencing we can quantitatively measure which genes are activated and their expression levels. By comparing the gene expression levels of normal and mutated cells, we can identify differences that might reveal which genes are causing these mutations. These gene expression patterns can also serve as indicators to detect and differentiate similar diseases such as AD and PSP.\n",
    "\n",
    "\n",
    "<span id=\"fn1\"> <sup>1</sup> https://www.nature.com/articles/sdata201689</span>\n",
    "\n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7b81483-4bac-45ad-904c-514b80965896",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "### Section 2: Data importing/preprocessing -  [10 points]\n",
    "\n",
    "- Load the gene expression dataset and the labels into pandas dataframe. \n",
    "- Retain only 'Diagnosis' and 'ID' columns in the labels dataframe.\n",
    "- Combine gene_exp and labels dataframes based on 'ID'. \n",
    "- Remove samples containing labels other than Alzheimer's ('AD') or Progressive supranuclear palsy ('PSP').\n",
    "- Encode textual labels to numbers. For example 'AD' to 0 and 'PSP' to 1.\n",
    "\n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "870a0220-08e6-467f-ab00-90dff1c36d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# do not modify this\n",
    "np.random.seed(89802024)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "149be304-4874-4965-afb8-da43f4afd8af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <th>ENSG00000240361</th>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <th>ENSG00000233750</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <th>ENSG00000210176</th>\n",
       "      <th>ENSG00000210184</th>\n",
       "      <th>ENSG00000210191</th>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <th>ENSG00000210194</th>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <th>ENSG00000210195</th>\n",
       "      <th>ENSG00000210196</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11344_TCX</td>\n",
       "      <td>-2.168275</td>\n",
       "      <td>2.867350</td>\n",
       "      <td>-2.235659</td>\n",
       "      <td>-2.878004</td>\n",
       "      <td>-2.878004</td>\n",
       "      <td>-2.921482</td>\n",
       "      <td>0.972981</td>\n",
       "      <td>2.589508</td>\n",
       "      <td>2.049147</td>\n",
       "      <td>...</td>\n",
       "      <td>15.177233</td>\n",
       "      <td>8.620554</td>\n",
       "      <td>8.415900</td>\n",
       "      <td>9.195871</td>\n",
       "      <td>13.407342</td>\n",
       "      <td>13.089514</td>\n",
       "      <td>10.035159</td>\n",
       "      <td>14.350876</td>\n",
       "      <td>8.401389</td>\n",
       "      <td>9.124693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11316_TCX</td>\n",
       "      <td>-2.667791</td>\n",
       "      <td>2.919402</td>\n",
       "      <td>-3.118351</td>\n",
       "      <td>-2.963895</td>\n",
       "      <td>-2.963895</td>\n",
       "      <td>-2.963895</td>\n",
       "      <td>1.602722</td>\n",
       "      <td>3.196802</td>\n",
       "      <td>2.830484</td>\n",
       "      <td>...</td>\n",
       "      <td>15.177233</td>\n",
       "      <td>11.611706</td>\n",
       "      <td>11.478740</td>\n",
       "      <td>12.284014</td>\n",
       "      <td>13.527804</td>\n",
       "      <td>13.350245</td>\n",
       "      <td>12.613126</td>\n",
       "      <td>14.240878</td>\n",
       "      <td>10.760811</td>\n",
       "      <td>12.021301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11431_TCX</td>\n",
       "      <td>-3.391138</td>\n",
       "      <td>3.153625</td>\n",
       "      <td>-2.106056</td>\n",
       "      <td>-4.424961</td>\n",
       "      <td>-4.418319</td>\n",
       "      <td>-4.540941</td>\n",
       "      <td>1.453881</td>\n",
       "      <td>2.986039</td>\n",
       "      <td>2.836847</td>\n",
       "      <td>...</td>\n",
       "      <td>15.177233</td>\n",
       "      <td>10.502635</td>\n",
       "      <td>10.133659</td>\n",
       "      <td>11.504803</td>\n",
       "      <td>13.407342</td>\n",
       "      <td>13.138792</td>\n",
       "      <td>11.990068</td>\n",
       "      <td>14.472362</td>\n",
       "      <td>9.848741</td>\n",
       "      <td>10.852588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11341_TCX</td>\n",
       "      <td>-2.801722</td>\n",
       "      <td>3.269001</td>\n",
       "      <td>-1.133302</td>\n",
       "      <td>-3.118351</td>\n",
       "      <td>-3.118351</td>\n",
       "      <td>-3.175145</td>\n",
       "      <td>2.442415</td>\n",
       "      <td>3.961310</td>\n",
       "      <td>2.803290</td>\n",
       "      <td>...</td>\n",
       "      <td>14.764855</td>\n",
       "      <td>9.920577</td>\n",
       "      <td>10.133659</td>\n",
       "      <td>11.255892</td>\n",
       "      <td>13.466460</td>\n",
       "      <td>13.350245</td>\n",
       "      <td>12.182661</td>\n",
       "      <td>14.240878</td>\n",
       "      <td>9.637298</td>\n",
       "      <td>9.904126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11289_TCX</td>\n",
       "      <td>-0.162419</td>\n",
       "      <td>5.284041</td>\n",
       "      <td>0.250976</td>\n",
       "      <td>-2.218957</td>\n",
       "      <td>-2.218957</td>\n",
       "      <td>-2.328570</td>\n",
       "      <td>2.065006</td>\n",
       "      <td>3.218117</td>\n",
       "      <td>2.857921</td>\n",
       "      <td>...</td>\n",
       "      <td>14.764855</td>\n",
       "      <td>11.780232</td>\n",
       "      <td>11.209200</td>\n",
       "      <td>12.021301</td>\n",
       "      <td>13.350245</td>\n",
       "      <td>13.189390</td>\n",
       "      <td>12.353690</td>\n",
       "      <td>14.350876</td>\n",
       "      <td>11.722941</td>\n",
       "      <td>11.959163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 55757 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID  ENSG00000223972  ENSG00000227232  ENSG00000243485  \\\n",
       "0  11344_TCX        -2.168275         2.867350        -2.235659   \n",
       "1  11316_TCX        -2.667791         2.919402        -3.118351   \n",
       "2  11431_TCX        -3.391138         3.153625        -2.106056   \n",
       "3  11341_TCX        -2.801722         3.269001        -1.133302   \n",
       "4  11289_TCX        -0.162419         5.284041         0.250976   \n",
       "\n",
       "   ENSG00000268020  ENSG00000240361  ENSG00000186092  ENSG00000238009  \\\n",
       "0        -2.878004        -2.878004        -2.921482         0.972981   \n",
       "1        -2.963895        -2.963895        -2.963895         1.602722   \n",
       "2        -4.424961        -4.418319        -4.540941         1.453881   \n",
       "3        -3.118351        -3.118351        -3.175145         2.442415   \n",
       "4        -2.218957        -2.218957        -2.328570         2.065006   \n",
       "\n",
       "   ENSG00000239945  ENSG00000233750  ...  ENSG00000198886  ENSG00000210176  \\\n",
       "0         2.589508         2.049147  ...        15.177233         8.620554   \n",
       "1         3.196802         2.830484  ...        15.177233        11.611706   \n",
       "2         2.986039         2.836847  ...        15.177233        10.502635   \n",
       "3         3.961310         2.803290  ...        14.764855         9.920577   \n",
       "4         3.218117         2.857921  ...        14.764855        11.780232   \n",
       "\n",
       "   ENSG00000210184  ENSG00000210191  ENSG00000198786  ENSG00000198695  \\\n",
       "0         8.415900         9.195871        13.407342        13.089514   \n",
       "1        11.478740        12.284014        13.527804        13.350245   \n",
       "2        10.133659        11.504803        13.407342        13.138792   \n",
       "3        10.133659        11.255892        13.466460        13.350245   \n",
       "4        11.209200        12.021301        13.350245        13.189390   \n",
       "\n",
       "   ENSG00000210194  ENSG00000198727  ENSG00000210195  ENSG00000210196  \n",
       "0        10.035159        14.350876         8.401389         9.124693  \n",
       "1        12.613126        14.240878        10.760811        12.021301  \n",
       "2        11.990068        14.472362         9.848741        10.852588  \n",
       "3        12.182661        14.240878         9.637298         9.904126  \n",
       "4        12.353690        14.350876        11.722941        11.959163  \n",
       "\n",
       "[5 rows x 55757 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the gene expression dataset (normalized_gene_expression.txt) into a dataframe\n",
    "\n",
    "gene_exp = pd.read_csv('normalized_gene_expression.txt', sep='\\t')\n",
    "gene_exp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5002922b-0af6-468a-866f-5ab292b5996f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>RIN</th>\n",
       "      <th>3primeBias</th>\n",
       "      <th>Sex</th>\n",
       "      <th>AgeAtDeath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1005_TCX</td>\n",
       "      <td>AD</td>\n",
       "      <td>8.6</td>\n",
       "      <td>1.333306</td>\n",
       "      <td>0</td>\n",
       "      <td>90_or_above</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1010_TCX</td>\n",
       "      <td>AD</td>\n",
       "      <td>8.3</td>\n",
       "      <td>1.593823</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1015_TCX</td>\n",
       "      <td>AD</td>\n",
       "      <td>8.7</td>\n",
       "      <td>1.671713</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1019_TCX</td>\n",
       "      <td>AD</td>\n",
       "      <td>7.8</td>\n",
       "      <td>1.402270</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1029_TCX</td>\n",
       "      <td>AD</td>\n",
       "      <td>9.7</td>\n",
       "      <td>1.599649</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID Diagnosis  RIN  3primeBias  Sex   AgeAtDeath\n",
       "0  1005_TCX        AD  8.6    1.333306    0  90_or_above\n",
       "1  1010_TCX        AD  8.3    1.593823    1           63\n",
       "2  1015_TCX        AD  8.7    1.671713    1           77\n",
       "3  1019_TCX        AD  7.8    1.402270    0           86\n",
       "4  1029_TCX        AD  9.7    1.599649    0           69"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the labels for the dataset from labels.csv file\n",
    "\n",
    "labels = pd.read_csv('labels.csv')\n",
    "labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0d535de7-08fd-4f8b-89f6-6d20f65ce0e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Diagnosis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1005_TCX</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1010_TCX</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1015_TCX</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1019_TCX</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1029_TCX</td>\n",
       "      <td>AD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID Diagnosis\n",
       "0  1005_TCX        AD\n",
       "1  1010_TCX        AD\n",
       "2  1015_TCX        AD\n",
       "3  1019_TCX        AD\n",
       "4  1029_TCX        AD"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop unnecessary columns in labels dataframe\n",
    "\n",
    "labels = labels.drop(columns=['RIN','3primeBias','Sex','AgeAtDeath'])\n",
    "labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "61d1193e-5c87-41d6-914a-83048b9c779c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Diagnosis\n",
       "PSP                 82\n",
       "AD                  80\n",
       "Control             76\n",
       "Pathologic Aging    30\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge both dataframes based on column 'ID' \n",
    "\n",
    "df = pd.merge(gene_exp, labels, on='ID')\n",
    "df['Diagnosis'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b648931-388c-44ec-b0c3-be279ea69e80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retain only samples having labels of 'AD' or 'PSP'\n",
    "\n",
    "df = \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157c3fcc-ffb1-4e4f-aba8-c797853779ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate the merged dataframe into features and labels. Remove 'ID' column\n",
    "\n",
    "X = # features\n",
    "y = # labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "583a3376-204c-4f56-af26-68694875554a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Encode textual labels in 'y' to numbers\n",
    "\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "y = \n",
    "\n",
    "assert(np.all((y == 0) | (y == 1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e7f1144-ee75-4468-a74e-798cdc145d4f",
   "metadata": {},
   "source": [
    "--------------------------\n",
    "### Section 3 - Feature selection and dimensionality reduction -  [15 points]\n",
    "\n",
    "On analyzing the shape of the gene expression dataframe we can see that the number of features (p) is much larger than the number of observations (N). This is known as the p >> N problem, and is commonly seen in genomics and healthcare datasets.\n",
    "To mitigate this problem, we will compare the effects of different feature selection and dimensionality reduction techniques so that we have a more compact feature space. These methods include:\n",
    "- Randomly selecting 100 features.\n",
    "- Selecting the top 1% of most variant features.\n",
    "- Performing PCA and select principal components that account for 95% of the total explained variance.\n",
    "    - Plot the first two principal components of X_projected, coloring data points by their labels. Explain if the data is linearly separable in this plot.\n",
    "\n",
    "  \n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef64079-1149-471e-9b9a-49460b5acbef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly select 100 features\n",
    "\n",
    "X_rand = \n",
    "\n",
    "assert X_rand.shape == (162, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7de16b56-3f6e-4834-b6ab-269571d36ffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the top 1% of the features with the most variance (hint: np.var)\n",
    "\n",
    "X_var = \n",
    "\n",
    "X_var.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7d2c821-0277-426b-ae8e-fb8884f1dd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use PCA and select only the components contributing to 95% of the total variance\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "X_projected = \n",
    "\n",
    "X_projected.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095cde99-35d7-4a14-ac5e-824571c5b883",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter plot of the first two principal components. \n",
    "\n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "plt.scatter(???) \n",
    "plt.scatter(???) \n",
    "plt.title('PCA: First Two Principal Components') \n",
    "plt.xlabel('PC 1') \n",
    "plt.ylabel('PC 2') \n",
    "plt.legend() \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6e3932c-58c8-490c-bb28-32e8a8242908",
   "metadata": {},
   "source": [
    "#### Q1: Is the dataset linearly separable?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749084a2-aeea-49ab-a9f7-4de0751baa04",
   "metadata": {},
   "source": [
    "[ANSWER HERE]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1adefb2e-8582-446b-b1b1-a21b62f28884",
   "metadata": {},
   "source": [
    "### Section 4: SVM classifier  - [20 points]\n",
    " \n",
    "\n",
    "- Complete the 'evaluate_classifier' function to implement a SVM classifier. Use GridSearchCV to search for the optimal kernel and parameter C (C determines the margin width).\n",
    "- Use 10-fold StratifiedKFold for cross-validation (CV) when searching. Set scoring to 'accuracy'. StratifiedKFold will divide our dataset into 10 sets, the model will be trained on 9 sets and tested on 1 set. This results in 10 models, each fitted on different combinations of sets. \n",
    "- Print the best parameters and score found for each of X_rand, X_var, and X_projected after cross validation.\n",
    "\n",
    "NOTE: We will not be doing a train-val-test split of the dataset in this assignment because of the small sample size. We will do a cross-validated train-test split for evaluations.\n",
    "\n",
    "\n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8645310c-71b7-4fb4-8ea9-e9a9ce7175cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import svm\n",
    "from sklearn.model_selection import StratifiedKFold, GridSearchCV\n",
    "\n",
    "\n",
    "def evaluate_classifier(clf, X, y, scoring, parameters, n_splits=10):\n",
    "    # Create StratifiedKFold validation\n",
    "\n",
    "    # Grid search for the two parameters (kernel, regularization parameter (inverse of margin))\n",
    "\n",
    "    # Print the best parameters and score\n",
    "\n",
    "    return grid_search\n",
    "\n",
    "\n",
    "parameters = {'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'C': [0.01, 0.1, 1, 10]}\n",
    "scoring = 'accuracy'\n",
    "\n",
    "# Intiate a SVM classifier\n",
    "clf = \n",
    "\n",
    "evaluate_classifier(clf, X_rand, y, scoring, parameters)\n",
    "evaluate_classifier(clf, X_var, y, scoring, parameters)\n",
    "evaluate_classifier(clf, X_projected, y, scoring, parameters)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb947f7-62f4-4021-9de0-a7363261a23d",
   "metadata": {},
   "source": [
    "#### Q2: Is `X_projected` performing better than the other two?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d327252b-0ab7-4fea-9af0-7a5477144826",
   "metadata": {},
   "source": [
    "[ANSWER HERE]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d49ac99-7277-42f1-af89-7d10d50d8353",
   "metadata": {},
   "source": [
    "### Section 5: Logistic regression - [30 points]\n",
    "\n",
    "We explored three basic methods for feature selection. Another approach for having fewer features is regularization. In lasso regularization, as we increase the parameter (it is called 'alpha' in sklearn), more and more coefficients of our features are shrunk to zero.\n",
    "\n",
    "- Build logistic regression models using lasso regularization (L1 penalty) with varying alpha values.  \n",
    "- We'll perform cross-validation using the 'cross_validate' function to measure the performance of models. We will again use StratifiedKFold with 10 splits, but remember to get the splits on the original dataframe 'X'. \n",
    "    - Set scoring to 'accuracy'.\n",
    "    - Record the mean test accuracy across CV splits for each alpha.\n",
    "    - Record the mean number of non-zero coefficients across CV splits for each alpha.\n",
    "- Repeat the above process but use ridge regression (l2 penalty) in logistic regression.\n",
    "- Plot the number of non-zero feature coefficients recorded in both lasso and ridge regression versus alpha.  \n",
    "- Plot cv accuracy recorded for both penalties versus alpha. Write two lines explaining why accuracy decreases as we increase alpha in lasso regression?\n",
    "\n",
    "\n",
    "Note: It may take awhile to run. Also, you can ignore any possible messages regarding convergence.\n",
    "\n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576ff578-4796-4bb5-bac4-99206cec1dd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import cross_validate\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "\n",
    "alphas = np.array([0.1, 1, 10, 100, 1000])\n",
    "lasso_cv_acc = []\n",
    "lasso_non_zero_coeff = []\n",
    "ridge_non_zero_coeff = []\n",
    "ridge_cv_acc = []\n",
    "\n",
    "# Create StratifiedKFold validation of 10 splits on 'X'\n",
    "\n",
    "for a in alphas:\n",
    "    # Instantiate a logistic regression object. Use 'saga' solver and 'l1' penalty. \n",
    "    \n",
    "    # Read on how parameter'C' and 'alpha' are related in logistic regression\n",
    "\n",
    "    # Perform cross validation as described above.\n",
    "\n",
    "    # Record mean of test_score of all 10 models.\n",
    "    \n",
    "    # Record mean of number of non zero coeefeicitens of all 10 models\n",
    "\n",
    "for a in alphas:\n",
    "\n",
    "    # Repeat the above but with 'l2' penalty \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "178e4aec-4754-4e6f-957a-0cd31b75f60e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot log of alpha vs number of non zero features. \n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "plt.xlabel('Log alpha')\n",
    "plt.ylabel('Number of non-zero features')\n",
    "plt.plot(np.log(alphas), ???, label='Lasso')\n",
    "plt.plot(np.log(alphas), ???, label='Ridge')\n",
    "plt.legend(loc=\"lower left\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac75a83-9c83-4c5f-8906-cb6eb979b10b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot log of alpha vs mean CV test accuracy. \n",
    "\n",
    "plt.xlabel('Log alpha')\n",
    "plt.ylabel('CV test accuracy')\n",
    "plt.plot(np.log(alphas), ???, label='Lasso')\n",
    "plt.plot(np.log(alphas), ???, label='Ridge')\n",
    "plt.legend(loc=\"lower left\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927c11d1-6c28-4be8-9da1-ab77b91422f0",
   "metadata": {},
   "source": [
    "#### Q3: Write a brief explanation for why the accuracy is decreasing in Lasso.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3640950f-3cd7-4960-b52b-503f301bac4c",
   "metadata": {},
   "source": [
    "[ANSWER HERE]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bba4b84-ad92-4e7a-88d1-2ded77dc9737",
   "metadata": {},
   "source": [
    "### Section 6: Random forest  - [25 points]\n",
    " \n",
    "In this section, we will implement a random forest algorithm on X_projected (i.e., the data reduced using PCA). We will vary the number of trees and features considered when splitting nodes. \n",
    "\n",
    "-  Build random forest models with varying tree counts and features to consider when looking for the split.\n",
    "-  Perform similar cross-validate with stratified folds, as in Section 5..\n",
    "-  Record accuracies obtained for the test set for each hyperparameter combination.\n",
    "-  Plot test accuracy vs tree count for each #features (m, m//2, sqrt(m)) we are using.\n",
    "-  Repeat the above using just the first 25 components of X_projected (you may create a new cell(s) for this).\n",
    "\n",
    "Note: Don't forget the last step. \n",
    "\n",
    "--------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19970263-0e64-4533-a5c9-aa8fdae60c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "X_rf = X_projected\n",
    "\n",
    "num_trees = [1, 50, 100, 200, 300, 500, 1000]\n",
    "m =  X_rf.shape[1]\n",
    "max_features = [ m, m//2, 'sqrt']\n",
    "test_accs = []\n",
    "\n",
    "# Create StratifiedKFold validation\n",
    "\n",
    "for features in max_features:\n",
    "    for tree in num_trees:\n",
    "        # Instantiate a random forest object. Set max_depth=5. \n",
    "\n",
    "        # Perform cross validation\n",
    "\n",
    "        # Record the mean of CV test scores \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2579cd1-25c3-4182-8d9a-3b4017aff834",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot CV test accuracy vs number of trees \n",
    "\n",
    "plt.xlabel('Number of Trees')\n",
    "plt.ylabel('CV test accuracy')\n",
    "for i, max_feat in enumerate(max_features):\n",
    "    plt.plot(num_trees, ??? , label=\"max_features = \"+str(max_feat))\n",
    "    \n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7b17351-894f-47d7-8836-f10b342b0dc8",
   "metadata": {},
   "source": [
    "#### Q4: What are your observations from the plots above?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a41823ea-88e5-4275-86bd-601ba697a9c3",
   "metadata": {},
   "source": [
    "[ANSWER HERE]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "748a1960-b5ad-449c-b4d6-cdb6f0f698c7",
   "metadata": {},
   "source": [
    "### 1 minute feedback \n",
    "\n",
    "Your feedback is valuable! Please take some time to fill out the assignment feedback form using the below link. Thank you.\n",
    "\n",
    "This form is anonymous. \n",
    "\n",
    "https://forms.gle/S6kVsKaiL7ciKHWw9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b054df59-8944-4cec-a396-c8c6f3cca073",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
